### ISSUE TITLE
Glowroot on kubernetes - Very slow performance

### ISSUE THREAD

**Issue by ejbp** [glowroot/glowroot#801]
I've tried this great tool on my dev server as standalone. It worked perfectly. Though, after spending some time installing it on kubernetes with glowroot-central I noticed that the performance went down (very bad performance, in fact).

I tried multi dir in first place because I wanted to have same configuration for all but I went with some lock issues as I using same agent id for different instances of same application, which makes sense because documentation requires agent id to be unique. Though, I decided to try with a box for each instance, meaning to have a different glowroot folder for each instance/pod/agent.

I read that the agent id should be unique but when not in a single box does it requires to be unique too for different instances/pods of same application?

My goal is to have only one entry per application on the glowroot-ui and not one per each instance/pod of each app.

Could the performance issue a consequence of using same agent-id on different instances/pods for the same app? Please not that each pod has it's own "disk" meaning that the reason is not a lock in the local agent.   

I'm running cassandra on the kubernetes too and I gave a maximum of 1 cpu and 2GB memory for Kassandra. For the glowroot central I gave only 512MB max memory usage.

Cassandra is writing on a persistent efs aws storage, so it's a remote disk. Agents are using local disk (not persistent).

An example:
 - Before adding glowroot as an agent,  my war file takes 30seconds to deploy.
 - After adding  glowroot as an agent, my war file takes 51seconds to deploy
 - Complex queries takes 3x to 4x more 

I had to disable for now but I would love to use this tool in production. Thanks in advance.

**Comment by alex-klimov** [comment]
Hello, we are working on recommended setup of Grlowroot in Kubernetes.

Specifically for your issue:
- When talking about the performance - do you mean the performance of the application start-up or something else?
- Do you see the same regression when running outside of Kubernetes?
- Could you please try Glowroot agent to apply to your application on your dev box - would you observe the same performance issue?

**Comment by ejbp** [comment]
Hi,

> When talking about the performance - do you mean the performance of the application start-up or something else?

Startup and during runtime. Endpoints with queries, for example, can take 3x to 4x more. I also tried to warm-up the jvm but still no success after it, meaning that the delay still persists.

> Do you see the same regression when running outside of Kubernetes? 

I've made some statistics tests outside kubernetes and I don't see any relevant difference neither during startup neither during requests. The point is that on this tests I'm using it as standalone and on kubernetes I tested with glowroot-central. I just remember that the issue may also be related with incompatibility with x-ray aws plugin (we only use it in production). I didn't remove the x-ray instrumentation while testing glowroot in production. Do you think it could be related with that?

> Could you please try Glowroot agent to apply to your application on your dev box - would you observe the same performance issue?

I need to setup it first. I'll give a try and comeback to you.

**Comment by alex-klimov** [comment]
Yes, when debugging performance related problems - that is the best to test only one thing at a time.

Regarding x-ray instrumentation - yes, turning it off would be advisable to have only one Java agent present. Is there a need to keep it running?

**Comment by ejbp** [comment]
**x-ray**
On my case there's no need since I want to complete replace x-ray by glowroot. But I don't want to migrate until I'm completely sure that works.

**Comment by deepaklalkbgmail** [comment]
any updates on this?

**Comment by Never-lab** [comment]
Two separate topics:

**1) One UI entry for many pods** — `agent.id` must still be **unique per JVM**. To group them under one app in Central, use a hierarchical id, e.g. `MyApp::pod-a`, `MyApp::pod-b` (see how Glowroot builds rollup paths with `::`). Same id on two live agents is wrong even on different disks.

**2) K8s slowness** — Central **512 MB** + Cassandra **1 CPU / 2 GB on remote EFS** is far too small for many agents; expect deploy and query slowdowns (same capacity family as #893 / #715). Also drop any second javaagent (x-ray) while measuring. Raise C*/Central resources, local SSD for Cassandra, unique hierarchical agent ids, current 0.14.x.

@nowheresly — close as support / capacity + config, or keep if you want a K8s sizing doc.
